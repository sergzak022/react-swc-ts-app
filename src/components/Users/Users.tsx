import {
  SyntheticEvent, useMemo, useState,
  useCallback,
} from 'react';

export interface NewUser {
  email: string;
  name: string;
}

export interface User extends NewUser {
  id: number;
}


export interface UsersProps {
  users: Array<User>
  onUserChange: (u: User) => void
  onUserCreate: (u: NewUser) => void
  onUserDelete: (id: number) => void
  selectedUser?: User
  onUserSelectBtnClick?: ( u: User ) => void
}

const EMPTY_NEW_USER = {email: '', name: ''};

export function Users (props: UsersProps) {
  const {
    users,
    onUserChange,
    onUserCreate,
    onUserDelete,
    selectedUser,
    onUserSelectBtnClick,
  } = props;

  const [newUser, setNewUser] = useState<NewUser>(EMPTY_NEW_USER);

  const onEmailChange = useCallback(
    function (e: SyntheticEvent<HTMLInputElement>, user: User) {
      const {currentTarget: {value}} = e;
      onUserChange({...user, email: value})
    },
    [onUserChange]
  )

  const onNameChange = useCallback(
    function (e: SyntheticEvent<HTMLInputElement>, user: User) {
      const {currentTarget: {value}} = e;
      onUserChange({...user, name: value})
    },
    [onUserChange]
  )

  const onCreate = useCallback(
    function (newUser: NewUser) {
      onUserCreate(newUser);
      setNewUser(EMPTY_NEW_USER);
    },
    [onUserCreate]
  )

  const usersList =  useMemo(() => {
    return users.map((user) => {
      const { id, email, name } = user;
      const isSelected = user === selectedUser

      return (
        <li data-testid="User" key={id}>
          <div>
            {isSelected && <span data-testid="User-selection">*</span>}
            <span data-testid="User-email">{email}</span>
            {' '} | {' '}
            <span data-testid="User-name">{name}</span>
          </div>
          <div>
            <label>
              Chage Email: 
              <input
                data-testid="User-email-input"
                onChange={(e) => onEmailChange(e, user)} value={email}
              />
            </label>
            <label>
              Chage name: 
              <input
                data-testid="User-name-input"
                onChange={(e) => onNameChange(e, user)} value={name}
              />
            </label>
            <label>
              <button
                data-testid="User-delete"
                onClick={() => onUserDelete(id)}
              >Delete</button>
              <button
                data-testid="User-select"
                onClick={
                  () =>
                    typeof onUserSelectBtnClick === 'function'
                      && onUserSelectBtnClick(user)
                }
              >Select</button>
            </label>
          </div>
        </li>
      )
    });
  }, [users, selectedUser, onEmailChange, onNameChange, onUserDelete, onUserSelectBtnClick])

  const { email, name } = newUser;

  return (
    <div>
      <div>
        <div>Create User</div>
        <label>
          <span>Email: </span>
          <input
            data-testid="User-new-email-input"
            onChange={({currentTarget: {value}}: SyntheticEvent<HTMLInputElement>) => setNewUser({email: value, name})}
            value={email}
          />
        </label>
        <label>
          <span>Name: </span>
          <input
            data-testid="User-new-name-input"
            onChange={({currentTarget: {value}}: SyntheticEvent<HTMLInputElement>) => setNewUser({email, name: value})}
            value={name}
          />
        </label>
        <label>
          <button
            data-testid="User-new-btn"
            disabled={email.length === 0 || name.length === 0}
            onClick={() => onCreate(newUser)}
          >Create</button>
        </label>
      </div>
      <ul data-testid="Users-list">
        {usersList}
      </ul>
    </div>
  )
}
