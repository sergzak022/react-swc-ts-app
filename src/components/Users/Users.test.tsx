import {
  render, screen, within, fireEvent
} from '@testing-library/react'
import '@testing-library/jest-dom'
import { Users, UsersProps } from './Users';

// TODO: use these constants in the component, so it easy to
// change values in the future
const USER_LIST_ITEM_ID = 'User'
const USER_EMAIL_ID = 'User-email'
const USER_NAME_ID = 'User-name'
const USER_EMAIL_INPUT_ID = 'User-email-input'
const USER_NAME_INPUT_ID = 'User-name-input'
const USER_DELETE_ID = 'User-delete'
const USER_SELECT_BTN_ID = 'User-select'
const USER_SELECTION_ID = 'User-selection'

const USER_NEW_EMAIL_INPUT_ID = 'User-new-email-input'
const USER_NEW_NAME_INPUT_ID = 'User-new-name-input'
const USER_NEW_BTN_ID = 'User-new-btn'


const getUserList = () => ([
  {id: 0, name: 'foo', email: 'foo@gmail.com'},
  {id: 1, name: 'bar', email: 'bar@gmail.com'},
  {id: 2, name: 'baz', email: 'baz@gmail.com'}
])

let props: UsersProps;

beforeEach(() => {
  const users = getUserList()

  props = {
    users,
    onUserChange: jest.fn(),
    onUserCreate: jest.fn(),
    onUserDelete: jest.fn(),
  }
})


test('should match snapshot', () => {
  const tree = render(<Users {...props} />)
  expect(tree).toMatchSnapshot()
})

test('should render non-empty list of users', async () => {
    render(<Users {...props} />)

    const usersList = screen.getAllByTestId(USER_LIST_ITEM_ID);
    expect(usersList.length).toEqual(3);
})

test('should render correct email for each users', async () => {
    render(<Users {...props} />)

    const usersList = screen.getAllByTestId(USER_LIST_ITEM_ID);
    const emails = usersList.map( userListEl => within(userListEl).getByTestId(USER_EMAIL_ID) )

    emails.forEach((email, idx) => expect(email).toHaveTextContent(props.users[idx].email));
})

test('should render correct name for each users', async () => {
    render(<Users {...props} />)

    const usersList = screen.getAllByTestId(USER_LIST_ITEM_ID);
    const emails = usersList.map( userListEl => within(userListEl).getByTestId(USER_NAME_ID) )

    emails.forEach((email, idx) => expect(email).toHaveTextContent(props.users[idx].name));
})

test('when change email should call onUserChange', async () => {
    render(<Users {...props} />)

    const emailInputList = screen.getAllByTestId(USER_EMAIL_INPUT_ID)

    emailInputList.forEach((emailInput, idx) => {
      fireEvent.change(emailInput, { target: { value: idx.toString() } })
    })

    const newUsers = props.users.map((user, idx) => ({
      ...user,
      email: idx.toString()
    }))

    const onUserChangeMock = props.onUserChange as jest.Mock;

    expect(props.onUserChange).toHaveBeenCalledTimes(3)

    onUserChangeMock.mock.calls.forEach(([ arg ], idx) => {
      expect(arg).toEqual(newUsers[idx])
    })
})

test('when change name should call onUserChange with right data', async () => {
    render(<Users {...props} />)

    const emailInputList = screen.getAllByTestId(USER_NAME_INPUT_ID)

    emailInputList.forEach((emailInput, idx) => {
      fireEvent.change(emailInput, { target: { value: idx.toString() } })
    })

    const newUsers = props.users.map((user, idx) => ({
      ...user,
      name: idx.toString()
    }))

    const onUserChangeMock = props.onUserChange as jest.Mock;

    expect(props.onUserChange).toHaveBeenCalledTimes(3)

    onUserChangeMock.mock.calls.forEach(([ arg ], idx) => {
      expect(arg).toEqual(newUsers[idx])
    })
})

test('when enter new user email, name and hit create button should call onUserCreate with right data', async () => {
  render(<Users {...props} />)

  const newUserEmailInput = screen.getByTestId(USER_NEW_EMAIL_INPUT_ID)
  const newUserNameInput = screen.getByTestId(USER_NEW_NAME_INPUT_ID)
  const newUserBtn = screen.getByTestId(USER_NEW_BTN_ID)

  const NewUser = {
    name: 'New Name',
    email: 'new@gmail.com',
  }

  fireEvent.change(newUserEmailInput, { target: { value: NewUser.email } })
  fireEvent.change(newUserNameInput, { target: { value: NewUser.name } })

  fireEvent.click(newUserBtn)

  expect(props.onUserCreate).toHaveBeenCalled();
  expect(props.onUserCreate).toHaveBeenCalledWith(NewUser);

})

test('when click delete btn should call onUserDelete', async () => {
  render(<Users {...props} />)

  const userDeleteBtns = screen.getAllByTestId(USER_DELETE_ID)

  userDeleteBtns.forEach((userDeleteBtn) => {
    fireEvent.click(userDeleteBtn)
  })

  expect(props.onUserDelete).toHaveBeenCalledTimes(props.users.length);
  expect(props.onUserDelete).toHaveBeenCalledWith(props.users[props.users.length - 1]?.id);

})

test('when selectedUser is set mark the selecte user row', () => {

  const selectedUserIdx = 1;
  const selectedUser = props.users[selectedUserIdx];

  const updatedProps = {
    ...props,
    selectedUser
  }

  render(<Users {...updatedProps} />)

  const userRows = screen.getAllByTestId(USER_LIST_ITEM_ID);

  const selectedUserRow = userRows.find( ( userRow ) => {
    return within(userRow).queryByTestId(USER_SELECTION_ID) != null
  })

  let selectedUserEmailEl,
      selectedUserNameEl,
      selectedRowIdx

  if (selectedUserRow) {
    selectedUserEmailEl = within(selectedUserRow).getByTestId(USER_EMAIL_ID)
    selectedUserNameEl = within(selectedUserRow).getByTestId(USER_NAME_ID)
    selectedRowIdx = userRows.indexOf(selectedUserRow)
  }

  expect(selectedRowIdx).toBe(selectedUserIdx)
  expect(selectedUserEmailEl).toHaveTextContent(selectedUser.email)
  expect(selectedUserNameEl).toHaveTextContent(selectedUser.name)

})

test('when selectedUser changes mark a differnt user row as selected', () => {
  // TODO
})

test('when select btn is clicked fire onUserSelectBtnClick', () => {

  const testUserIdx = 1;

  const updatedProps = {
    ...props,
    onUserSelectBtnClick: jest.fn(),
  }
  render(<Users {...updatedProps} />)

  const userRows = screen.getAllByTestId(USER_LIST_ITEM_ID)

  const secondUserSelectBtn = within(userRows[testUserIdx]).getByTestId(USER_SELECT_BTN_ID)

  fireEvent.click(secondUserSelectBtn)

  expect(updatedProps.onUserSelectBtnClick).toHaveBeenCalledWith(props.users[testUserIdx])
})
