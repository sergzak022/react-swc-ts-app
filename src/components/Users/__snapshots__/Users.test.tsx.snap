// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div>
          <div>
            Create User
          </div>
          <label>
            <span>
              Email: 
            </span>
            <input
              data-testid="User-new-email-input"
              value=""
            />
          </label>
          <label>
            <span>
              Name: 
            </span>
            <input
              data-testid="User-new-name-input"
              value=""
            />
          </label>
          <label>
            <button
              data-testid="User-new-btn"
              disabled=""
            >
              Create
            </button>
          </label>
        </div>
        <ul
          data-testid="Users-list"
        >
          <li
            data-testid="User"
          >
            <div>
              <span
                data-testid="User-email"
              >
                foo@gmail.com
              </span>
               
               | 
               
              <span
                data-testid="User-name"
              >
                foo
              </span>
            </div>
            <div>
              <label>
                Chage Email:
                <input
                  data-testid="User-email-input"
                  value="foo@gmail.com"
                />
              </label>
              <label>
                Chage name:
                <input
                  data-testid="User-name-input"
                  value="foo"
                />
              </label>
              <label>
                <button
                  data-testid="User-delete"
                >
                  Delete
                </button>
                <button
                  data-testid="User-select"
                >
                  Select
                </button>
              </label>
            </div>
          </li>
          <li
            data-testid="User"
          >
            <div>
              <span
                data-testid="User-email"
              >
                bar@gmail.com
              </span>
               
               | 
               
              <span
                data-testid="User-name"
              >
                bar
              </span>
            </div>
            <div>
              <label>
                Chage Email:
                <input
                  data-testid="User-email-input"
                  value="bar@gmail.com"
                />
              </label>
              <label>
                Chage name:
                <input
                  data-testid="User-name-input"
                  value="bar"
                />
              </label>
              <label>
                <button
                  data-testid="User-delete"
                >
                  Delete
                </button>
                <button
                  data-testid="User-select"
                >
                  Select
                </button>
              </label>
            </div>
          </li>
          <li
            data-testid="User"
          >
            <div>
              <span
                data-testid="User-email"
              >
                baz@gmail.com
              </span>
               
               | 
               
              <span
                data-testid="User-name"
              >
                baz
              </span>
            </div>
            <div>
              <label>
                Chage Email:
                <input
                  data-testid="User-email-input"
                  value="baz@gmail.com"
                />
              </label>
              <label>
                Chage name:
                <input
                  data-testid="User-name-input"
                  value="baz"
                />
              </label>
              <label>
                <button
                  data-testid="User-delete"
                >
                  Delete
                </button>
                <button
                  data-testid="User-select"
                >
                  Select
                </button>
              </label>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        <div>
          Create User
        </div>
        <label>
          <span>
            Email: 
          </span>
          <input
            data-testid="User-new-email-input"
            value=""
          />
        </label>
        <label>
          <span>
            Name: 
          </span>
          <input
            data-testid="User-new-name-input"
            value=""
          />
        </label>
        <label>
          <button
            data-testid="User-new-btn"
            disabled=""
          >
            Create
          </button>
        </label>
      </div>
      <ul
        data-testid="Users-list"
      >
        <li
          data-testid="User"
        >
          <div>
            <span
              data-testid="User-email"
            >
              foo@gmail.com
            </span>
             
             | 
             
            <span
              data-testid="User-name"
            >
              foo
            </span>
          </div>
          <div>
            <label>
              Chage Email:
              <input
                data-testid="User-email-input"
                value="foo@gmail.com"
              />
            </label>
            <label>
              Chage name:
              <input
                data-testid="User-name-input"
                value="foo"
              />
            </label>
            <label>
              <button
                data-testid="User-delete"
              >
                Delete
              </button>
              <button
                data-testid="User-select"
              >
                Select
              </button>
            </label>
          </div>
        </li>
        <li
          data-testid="User"
        >
          <div>
            <span
              data-testid="User-email"
            >
              bar@gmail.com
            </span>
             
             | 
             
            <span
              data-testid="User-name"
            >
              bar
            </span>
          </div>
          <div>
            <label>
              Chage Email:
              <input
                data-testid="User-email-input"
                value="bar@gmail.com"
              />
            </label>
            <label>
              Chage name:
              <input
                data-testid="User-name-input"
                value="bar"
              />
            </label>
            <label>
              <button
                data-testid="User-delete"
              >
                Delete
              </button>
              <button
                data-testid="User-select"
              >
                Select
              </button>
            </label>
          </div>
        </li>
        <li
          data-testid="User"
        >
          <div>
            <span
              data-testid="User-email"
            >
              baz@gmail.com
            </span>
             
             | 
             
            <span
              data-testid="User-name"
            >
              baz
            </span>
          </div>
          <div>
            <label>
              Chage Email:
              <input
                data-testid="User-email-input"
                value="baz@gmail.com"
              />
            </label>
            <label>
              Chage name:
              <input
                data-testid="User-name-input"
                value="baz"
              />
            </label>
            <label>
              <button
                data-testid="User-delete"
              >
                Delete
              </button>
              <button
                data-testid="User-select"
              >
                Select
              </button>
            </label>
          </div>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
